### PR TITLE
Improve operator_test with eventual consistency

### PR DESCRIPTION
**Pull request by erkanerol**
Signed-off-by: Erkan Erol <eerol@redhat.com>

**What this PR does / why we need it**:
These tests are reported as flaky:
https://storage.googleapis.com/kubevirt-prow/reports/flakefinder/kubevirt/kubevirt/flakefinder-2021-03-26-024h.html

and this PR aims to increase the stability of the test.

**Special notes for your reviewer**:

The logic in the current tests is as follow:
1) Change an operand
2) Expect that the change is reverted by the operator
3) Expect that the generation of operand is updated in the Kubevirt object
4) Expect that the generation remains the same and it achieves consistency.

The problem is that  `generation` of operand is saved to a variable just after step 2 and it is expected that it will remain the same. However, operator doesn't have to apply all expected changes at once and it can continue adapting the operand after reverting the change. Thefore, the generation after first observation of revert has not to be the final generation.

At the end of the day, the expected behavior,
- The change must be reverted
- The generation must be updated in Kubevirt

and it should be consistent as well. It is not important how many iterations happened in the generation or operand. 

This PR adapts the tests with this philosophy. 
**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Review by ksimon1 — COMMENTED** [review]
/lgtm

File: tests/operator_test.go [hunk 1]
```diff
@@ -912,24 +911,19 @@ spec:
...
 
-				return true
-			}, 120*time.Second, 5*time.Second).Should(BeTrue(), "waiting for deployment to revert to original state")
-
-			generation := vc.ObjectMeta.Generation
-			Eventually(func() int64 {
 				currentKV := tests.GetCurrentKv(virtClient)
-				return apply.GetExpectedGeneration(vc, currentKV.Status.Generations)
-			}, 60*time.Second, 5*time.Second).Should(Equal(generation), "reverted deployment generation should be set on KV resource")
+				return vc.ObjectMeta.Generation == apply.GetExpectedGeneration(vc, currentKV.Status.Generations)
```
[the remark this answers is not in the document]
Comment (phoracek) [reply tests/operator_test.go:915]: @erkanerol ^^

[the remark this answers is not in the document]
Comment (erkanerol) [reply tests/operator_test.go:915]: @phoracek The method does not return the generation of Kubevirt. It takes the generations array in the status field of Kubevirt and return the generation of given object in that array.

Shortly, it checks `whether the generation of the operand in Kubevirt's status is up-to-date`.

**Review by enp0s3 — COMMENTED** [review]
@erkanerol Thank you! changes looks good. However I have one concern about the intermediate generations that are seen here, @davidvossel @ashleyschuett can you please advice whether the reconciliation should be done in one operation?

File: tests/operator_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (phoracek) [reply tests/operator_test.go:915]: Oh, I see! Thanks a lot for the explanation

**Comment by ashleyschuett** [comment]
I disagree with the changes in this test. While this decreases the "flakeyness" of the test, I actually think this test failing is alerting us that there is something wrong with the reconciliation feature. 

Once the the resources is reverted back to it's original/required state there should be no more updates to the resource and the resource version should not change after this. I think enforcing this to be true is important. 

This test change instead enforces that the generation ID of a resource is the same as that of the KV resource, at the time it is being checked. It could be continuously hot looping and updating but we would not fail if these two resources stayed in sync. 

This change removes the ability for us to know when hot looping is occurring / when there is something going wrong with reconciliation. 

I have made another PR that should improve reconciliation https://github.com/kubevirt/kubevirt/pull/5470

**Comment by erkanerol** [comment]
> This change removes the ability for us to know when hot looping is occurring / when there is something going wrong with reconciliation.
> 
> I have made another PR that should improve reconciliation #5470

@ashleyschuett  I was aware of what we are losing but not 100% sure about whether we really need it or not. That is exactly why Igor wanted to get a review from you.  Good to see that you shared your concerns with us. I am closing this PR now.

**Pull request closed without merging**